thrust/detail/internal_functional.h

```diff
@@ -317,11 +318,11 @@ template<typename UnaryFunction>
   __thrust_exec_check_disable__
   template<typename Tuple>
   inline __host__ __device__
-  typename enable_if_non_const_reference_or_tuple_of_iterator_references<
-    typename thrust::tuple_element<1,Tuple>::type
-  >::type
-    operator()(Tuple t)
+  void operator()(Tuple t)
   {
+    THRUST_STATIC_ASSERT_MSG(is_non_const_reference<Tuple>::value ||
+                             is_tuple_of_iterator_references<Tuple>::value,
+      "Expected a non-const reference or thrust::detail::tuple_of_iterator_references");
```

Comment:
As discussed on telecon, the original conditional seems to be the right one.

We should also check whether the well-formedness of this call is used in any SFINAE condition on a public interface.

Reply:
Reverted this change, the real fix was to add

```
namespace thrust
{
namespace detail
{
// We need this type to pass as a non-const ref for unary_transform_functor
// to compile:
template <>
struct is_non_const_reference<only_set_when_expected_it> : thrust::true_type {};
}
}
```

to testing/copy.cu.